src/Storages/StorageDistributed.cpp

```diff
@@ -169,6 +170,113 @@ void replaceCurrentDatabaseFunction(ASTPtr & ast, const ContextPtr & context)
     for (auto & child : ast->children)
         replaceCurrentDatabaseFunction(child, context);
 }
+
+NameSet collectColumnsToCast(
+    const ColumnsDescription & metadata_columns,
+    const ColumnsDescription & actual_columns)
+{
+    NameSet names;
+    if (metadata_columns.empty() || actual_columns.empty())
+        return names;
+
+    for (const auto & expected_column : metadata_columns.getAllPhysical())
+    {
+        auto actual_column_opt = actual_columns.tryGetPhysical(expected_column.name);
+        if (!actual_column_opt)
+            continue;
+
+        if (!actual_column_opt->type->equals(*expected_column.type))
+            names.emplace(expected_column.name);
+    }
+
+    return names;
+}
+
+void applyHybridCastsToQueryTree(
+    QueryTreeNodePtr & query_tree,
+    const QueryTreeNodePtr & replacement_table_expression,
+    const ColumnsDescription & metadata_columns,
+    const NameSet & columns_to_cast,
+    const ColumnsDescription & actual_columns,
+    const ContextPtr & context)
+{
+    if (columns_to_cast.empty())
+        return;
+
+    class HybridCastVisitor : public InDepthQueryTreeVisitor<HybridCastVisitor>
```

Comment:
The code is simple, but still deserves documentation. Specially explaining on why this is needed